CloudFront CDN configuration and Varnish refactor

wjordan opened the pull request:
1. Introduces an application-specific HTTP cache-configuration hash that will be used for provisioning/configuring our HTTP cache layers moving forward.
2. Refactors the Varnish configuration to take advantage of the cache-configuration hash, which separates the application-specific configuration from the Varnish-specific VCL code. I've also taken this refactoring opportunity to upgrade Varnish to 4.0.3 which is the [current-supported release](https://www.varnish-cache.org/releases) (Varnish 3 was retired in April 2015).
3. Introduces a CloudFront CDN configuration script which uses the HTTP cache-configuration hash to create/update an environment-specific CloudFront distribution using the AWS SDK for Ruby.
4. Adds a set of Chef integration tests to the refactored `cdo-varnish` cookbook, to ensure that no regressions have been introduced during the refactoring.
5. `Rack::SslEnforcer` middleware will be added to all environments including production, to ensure that the HTTP `Strict-Transport-Security` response header is set from the application layer rather than from the HTTP-cache layer (this was previously being served by Varnish in non-development environments). HTTPS redirection is still handled by the HTTP-cache layer (CloudFront or Varnish) in non-development environments.

Aside from the middleware change, none of these changes affect the web-application code directly. The CloudFront CDN and Varnish cookbook are two separate pieces that are deployed independently of our webapp-deployment process. Steps for each rollout are documented below:
## CloudFront ([wiki](http://wiki.code.org/display/PROD/CDN+setup))

Setup steps for CloudFront CDN rollout across environments:

(In the below instructions, replace `ENV` with the environment name, e.g. `ENV-studio.code.org` is `staging-studio.code.org` in staging, with the exceptions of code.org and studio.code.org in production. `[num]`/`[abc]`/`[xyz]` are placeholders for arbitrary ID strings.)
- Run the `update_cloudfront` script with `RAILS_ENV` and AWS credentials provided in the environment. This will create/update the environment-specific CloudFront distribution based on the latest HTTP cache configuration in the codebase.
- In Route 53, add 2 new `A ALIAS` entries for the dashboard/pegasus ELB endpoints, that will serve as the origin domains for the CloudFront distributions:

```
ENV-dashboard.code.org A ALIAS dualstack.ENV-dashboard-[num].us-east-1.elb.amazonaws.com.
ENV-pegasus.code.org A ALIAS dualstack.ENV-pegasus-[num].us-east-1.elb.amazonaws.com.
```
- Test that the new hostnames serve the correct content when loading `ENV-dashboard.code.org` and `ENV-pegasus.code.org` directly.
- To switch on the CDN integration, update the `A` entries from the existing ELB endpoints to the corresponding CloudFront ones:

```
ENV-studio.code.org A ALIAS [abc].cloudfront.net.
ENV.code.org A ALIAS [xyz].cloudfront.net.
```

Note: When testing, it might be a good idea to set the TTL to a relatively low number of seconds, so that a rollback will be applied without too much delay if necessary.
- To rollback the CDN integration, simply update the `A` entries back to point directly to the ELB:

```
ENV-studio.code.org A ALIAS dualstack.ENV-dashboard-[num].us-east-1.elb.amazonaws.com.
ENV.code.org A ALIAS dualstack.ENV-pegasus-[num].us-east-1.elb.amazonaws.com.
```
## Varnish ([wiki](http://wiki.code.org/display/PROD/Chef-managed+Configuration+Changes))

For the Varnish update rollout I plan to test out a new versioned-cookbook deployment process:
- First, set the cookbook constraint to the existing cookbook version in all relevant environments one of these three ways:
  - Adjust the `cookbook_versions` property from `knife environment edit [ENV]`;
  - Adjust the "cookbook constraints" section in the `policy->environments` section of the Chef management console;
  - Running `knife cookbook upload [cookbook] -E [ENV]` with the existing cookbook present.
- Increase the cookbook version in `metadata.rb`, and make all cookbook changes.
- Upload the newly-versioned cookbook and update the environment-specific cookbook constraints one at a time, through one of the three ways listed above.

philbogle commented:
Could you link or copy the set up stepsin this pull request into the wiki to make sure they can be found?  (If there not finalized, it's fine if this doesn't happen right away.)

wjordan commented:
I've made the CloudFront rollout steps also available at [CDN Setup](http://wiki.code.org/display/PROD/CDN+setup). The Varnish rollout steps will be integrated into [Chef-managed Configuration Changes](http://wiki.code.org/display/PROD/Chef-managed+Configuration+Changes) once they have been tested and are finalized.

philbogle commented on cookbooks/cdo-varnish/README.md:
```diff
@@ -1,68 +1,39 @@
+  - `cookies`: A whitelist array of HTTP cookie keys to pass to the origin and include in the cache key.
+    To whitelist all cookies for the path, pass `'all'`.
+    To strip all cookies for the path, pass `'none'`.
+  - `proxy`: If specified, proxy all requests matching this path to the specified origin. (Currently either `'dashboard'` or `'pegasus'`)
```
Could you wrap these lines that don't fit in the github line length?

wjordan replied:
done

philbogle commented on cookbooks/cdo-varnish/libraries/helpers.rb:
```diff
@@ -0,0 +1,94 @@
+# if backend=true, call `bereq` instead of `req`.
+# if proxy=true, ignore extension-based behaviors (e.g., *.png).
+def paths_to_regex(behavior, backend=false, proxy=false)
+  req = backend ? 'bereq' : 'req'
```
The only part of the behavior access is the path;  how about just taking the path as a parameter?  (This will make the documentation more specifici and make this easier to test.)

wjordan replied:
done

wjordan replied:
done

philbogle commented on cookbooks/cdo-varnish/libraries/helpers.rb:
```diff
@@ -0,0 +1,94 @@
+  path_config = behavior['path']
+  path_config = [path_config] unless path_config.is_a?(Array)
+  extensions, paths = path_config.partition{ |path| path[0] == '*' }
+  elements = paths.map{|x|path_to_regex(x)}
```
This logic is complex enough that I think it needs unit tests; I'm think the current tests probably don't get full coverage.

wjordan replied:
The integration tests generally cover this logic, since they confirms that the generated VCL still compiles and behaves as expected. However I can add another layer of unit tests using [ChefSpec](https://github.com/sethvargo/chefspec) that can test these helper-methods more clearly and directly.

philbogle replied:
Thanks.  The integration tests are valuable but not sufficient; it's too difficult to cover the space of valid and invalid paths and domain names for example, whereas simple unit tests can easily and exhaustively do this by directly calling the methods through their simple API.

I am wondering whether you need ChefSpec; could you not factor these as libraries that are tested using our standard unit test framework.

philbogle commented on cookbooks/cdo-varnish/templates/default/default.vcl.erb:
```diff
@@ -1,231 +1,30 @@
-        if (is_supported(pl[i].lang))
-            RETURN_LANG(pl[i].lang);
+  <%= if_app(app, backend) %>
+    if (false) {
```
Why if(false)?  Better to remove code if it's dead.

wjordan replied:
The template-generated conditional-blocks here was pretty complicated. I've refactored this stuff into an `if_else` helper method that separates the conditional-block string-generation from the underlying code blocks.

wjordan commented:
A CircleCI integration-test run including the `cdo-varnish` tests is available [here](https://circleci.com/gh/wjordan/code-dot-org/53) (expand the `./test.sh` section to see the full log).

philbogle commented on cookbooks/cdo-varnish/libraries/helpers.rb:
```diff
@@ -0,0 +1,96 @@
+
+# Returns a regex-conditional string fragment based on the provided behavior.
+# In the 'proxy' section, ignore extension-based behaviors (e.g., *.png).
+def paths_to_regex(path_config, section)
```
Please document the valid path patterns and add unit tests that verify that allowed patterns are correctly translated and that disallowed patterns are rejected with an error.  The CloudFront document mentioned earlier in this file allows '*' and '?' anywhere in the path, but this appears to be more restrictive in what it can handle.

philbogle commented on cookbooks/cdo-varnish/libraries/helpers.rb:
```diff
@@ -0,0 +1,96 @@
+# Various Ruby helper methods to generate Varnish VCL from the base configuration.
+
+# Basic regex matcher for the query part of a URL.
+QUERY_REGEX = "(\\?[a-z0-9]+)?$"
```
There are additional characters allowed in a query string (e.g. '&' and '%'), does this need to be generalized to handle those?

wjordan replied:
Yes- this line was in the original VCL but a more general regex would be better. updated

philbogle commented on cookbooks/cdo-varnish/libraries/helpers.rb:
```diff
@@ -0,0 +1,96 @@
+QUERY_REGEX = "(\\?[a-z0-9]+)?$"
+
+def path_to_regex(path)
+  path = "^/#{path.sub(/^\//,'')}"
```
This might be easier to read as:

path = "/#{path}" unless path.start_with?('/')

philbogle commented on cookbooks/cdo-varnish/libraries/helpers.rb:
```diff
@@ -0,0 +1,117 @@
+
+def path_to_regex(path)
+  path = "^/#{path.sub(/^\//,'')}"
+  if path[-1] == '*'
```
This will generate incorrect/invalid regular expressions  or surprising behavior for a number of plausible path expressions that should be rejected and for which we would want to have unit tests.

Consider for example:
/images/_.jpg    (path with wildcarded filename)
/assets/_/images   (path with wildcard subexpression)
/path.1/2            (regular expression characters with special meaning like ., [, \, etc)
/a???/        (unsupported wildcard character supported in CDN paths)

wjordan replied:
Working on these unit tests as we speak.

philbogle commented on cookbooks/cdo-varnish/libraries/helpers.rb:
```diff
@@ -0,0 +1,117 @@
+# In the 'proxy' section, ignore extension-based behaviors (e.g., *.png).
+def paths_to_regex(path_config, section)
+  path_config = [path_config] unless path_config.is_a?(Array)
+  extensions, paths = path_config.partition{ |path| path[0] == '*' }
```
Consider also the invalid path  */images which will be considered as an extension by this logic.

philbogle commented on cookbooks/cdo-varnish/libraries/helpers.rb:
```diff
@@ -0,0 +1,117 @@
+
+# Returns the canonical hostname based on the current Chef node's name/environment.
+# Keep in sync with CDO.canonical_hostname in deployment.rb.
+def canonical_hostname(domain)
```
This should also be unit tested to make sure we cover all of the possible domains of interest.

philbogle commented:
High level comment:  I do like what you are doing to create a cache configuration hash which isn't tied specifically to Varnish and generating the Varnish config from that!

philbogle commented on cookbooks/cdo-varnish/README.md:
```diff
@@ -1,68 +1,50 @@
+- `behaviors`: Array of behaviors, evaluated in top-to-bottom order:
+  - `path`: [Path pattern](http://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/distribution-web-values-specify.html#DownloadDistValuesPathPattern)
+    to match this behavior against.
+    `*`-wildcards are allowed, either extension-wildcard `*.jpg` or
```
I think the link to the Amazon documentation is potentially misleading.  It includes example paths which I believe aren't supported here (e.g. /images/*.jpg and /a??.jpg)

wjordan commented:
@philbogle ready for the next review pass. I've refactored the various helper methods to reject invalid path patterns and added several unit tests to [shared/test/test_varnish_helpers.rb](https://github.com/code-dot-org/code-dot-org/pull/4207/files#diff-5fcdf436dc4538858904c459711574ef), among a few other minor tweaks.

philbogle commented:
Thanks, will take a look.

On Thu, Oct 1, 2015 at 5:50 PM, Will Jordan notifications@github.com
wrote:

> @philbogle https://github.com/philbogle ready for the next review pass.
> I've refactored the various helper methods to reject invalid path patterns
> and added several unit tests to shared/test/test_varnish_helpers.rb
> https://github.com/code-dot-org/code-dot-org/pull/4207/files#diff-5fcdf436dc4538858904c459711574ef,
> among a few other minor tweaks.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/code-dot-org/code-dot-org/pull/4207#issuecomment-144889264
> .

philbogle commented:
Looking now.

philbogle commented on cookbooks/cdo-varnish/libraries/helpers.rb:
```diff
@@ -0,0 +1,135 @@
+  # Maximum length
+  return false if path.length > 255
+  # Valid characters allowed
+  char = /[A-Za-z0-9_\-.$\/~"'@:+]/
```
This isn't the full set of valid characters allowed in an URL.  If you wanted you could tighten the documentation to be more restrictive if we don't care. 

http://stackoverflow.com/questions/1547899/which-characters-make-a-url-invalid
"n general URIs as defined by RFC 3986 (see Section 2: Characters) may contain any of the following characters: ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~:/?#[]@!$&'()*+,;="

wjordan commented:
FYI, I've added a `behavior_for_path` helper method, which will be used by a cookie-whitelisting Rack middleware I'm currently working on implementing. This middleware will read the same HttpCache configuration to simulate cookie-whitelisting behavior in development.

philbogle commented on cookbooks/cdo-varnish/libraries/http_cache.rb:
```diff
@@ -0,0 +1,95 @@
+# This hash provides extra application-specific configuration for whitelisting specific request headers and
+# cookies based on the request path.
+  def self.config(session_key, storage_id)
+    all_cookies = [
```
How about calling this hoc_cookies or whitelisted_cookies?

all_cookies is confusing when I read it below because it doesn't mean _all_ cookies, it mean's certain whilisted cookies.

wjordan replied:
sounds good!

philbogle commented on cookbooks/cdo-varnish/libraries/http_cache.rb:
```diff
@@ -0,0 +1,95 @@
+#
+# See cdo-varnish/README.md for more information on the configuration format.
+class HttpCache
+  STATIC_ASSETS = {
```
It's dangerous for correctness to strip all cookies based solely on extensions.

There are some assets (e.g. JPG ones) whose content depends on session state, so in general its not safe to blindly cache based on extension.  Brent has already made a VCL config change based on this;  how can we allow similar flexibility in this DSL?

Rather than an extremely complicated config, it would be nice if we had some simple path conventions where static and non-static (i.e not publically cacheable) assets go so that the rules here became simple path matches. I see that you've added a comment to that effect below.

For purposs of this PR, it's probably enough to make sure that you handle the exception Brent added to the VCL, we can work on the other broader issues in follow on PRs.

wjordan replied:
The newly-added exception has already been added to this PR (see 91e28c0e9ad8ddbd8bd022883f832cb26b63fc9f). 

Though this PR was designed as a straightforward translation of the existing VCL behaviors (which included the extension-based stripping as listed here), I agree that the broader discussion around a simpler set of cache-behavior path conventions is worth pursuing in follow-on PRs (which this simplified cache-configuration will hopefully make more feasible).

philbogle replied:
Could clarify in the documentation what happens when there are multiple matching behaviors?  Although it says they are evaluated in top-to-bottom order, it doesn't say that evaluation stops with the first matching path.  In the configuration snippet below, one might expect that the STATIC_ASSETS pattern would strip the cookies after the v3/assets behavior is run.
-        behaviors: [
-          {
-            path: 'v3/assets/*',
-            headers: ALL_HEADERS,
-            cookies: whitelisted_cookies
-          },
-          ...
-          STATIC_ASSETS,

wjordan replied:
Evaluation stops with the first matching path. I'll make this explicit in the documentation.

philbogle commented on cookbooks/cdo-varnish/libraries/http_cache.rb:
```diff
@@ -0,0 +1,100 @@
+
+# Host header is needed for Varnish to delegate to pegasus/dashboard depending on the host.
+# Accept-Language header is needed to cache language-specific pages.
+  ALL_HEADERS = %w(Host Accept-Language)
```
Could you please document what this means?  ALL_HEADERS is again confusing here since it isn't all headers but only a few headers.

Are these really the only headers passed on to the Rails frontend, or am I misinterpreting what they mean?  It seems like ALL_HEADERS is used for a lot of paths below.  There are lots of important request headers that are being stripped here that I would be surprised not to find in Rails controllers (HOST, Cookie, If-Modified-Since, etc) which make me think I must be misinterpreting things.  

This is another place where philosophically I wish we could list headers that we've analyzed and know are safe to strip, versus listed a small set which we think we need.  We really need to be running Varnish in dev to have a better chance of finding issues caused by the optimization of potentially unsafe stripping of headers we don't know about.

wjordan replied:
Yeah, the header-stripping part of the cache-configuration is still a bit unfinished- I need to do another pass at the documentation to make it clear what exactly the `header` field is used for in the current implementation, and what's still TODO. (Currently, `header` is only passed to the CloudFront distribution config but does not yet affect the Varnish configuration.)

For reference, here is the list of common HTTP request headers that CloudFront will forward to the origin without any additional configuration necessary, which should cover most typical HTTP usages: [HTTP Request Headers and CloudFront Behavior](http://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/RequestAndResponseBehaviorCustomOrigin.html#request-custom-headers-behavior)

There is an option to pass through all headers, which also disables caching entirely for the path (the syntax for this is `[*]` in the config hash, as documented), which we could use for paths where we worry about potentially unsafe stripping of headers. CloudFront does not support passing-through all but a specified list of headers to the origin.

The 'proxy' field is also currently partially-implemented, in the reverse direction- it is implemented in Varnish and the current configuration preserves the existing behavior, but is not yet implemented in CloudFront. I will need to finish the CloudFront implementation before we would be able to remove Varnish from our stack.

wjordan replied:
After looking at the CloudFront [documentation](http://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/RequestAndResponseBehaviorCustomOrigin.html#request-custom-headers-behavior) further and running some tests, I can clarify that any headers NOT listed in that table are considered "Customer-defined headers", which are forwarded to the origin even when a custom header is not included in the cache key.

So CloudFront only removes/changes the headers it explicitly mentions in the table (such as `Accept`, `Accept-Language`, and `User-Agent`), and leaves the rest untouched.

philbogle commented:
Here's a thought- we are going to need to exhaustively test HOC to make sure we hold up to load and that the desired public aching is happening in all cases (including in many otherwise dynamic pages which don't match the simple file extension list of things to cache.)   

The near future (though not this PR) is a good time to make the configuration simpler and safer where possible, because we have lots of time to test and correct anything that's not quite right;  the closer we get to HOC the more afraid we'll be to change things and the more complexity bits us.

This doesn't need to happen in this PR which should just preserve existing behavior but something to think of as we continue to evolve the conifguration.

philbogle commented on cookbooks/cdo-varnish/libraries/helpers.rb:
```diff
@@ -0,0 +1,148 @@
+  items.each_with_index do |item, i|
+    condition = conditional.call(item)
+    next if condition.to_s == 'false'
+    _buf << "#{i != 0 ? 'else ' : ''}#{condition && "if (#{condition}) "}{\n"
```
This is a bit dense.  Is the following equivalent?
buf << 'else ' if i != 0
buf << "if (#{condition}) "}{\n" if condition

philbogle commented on cookbooks/cdo-varnish/templates/default/accept-language.vcl.erb:
```diff
@@ -0,0 +1,188 @@
+#define LANG_LIST_SIZE 16
+#define HDR_MAXLEN 256
+#define LANG_MAXLEN 8
+#define RETURN_LANG(x) { \
```
I think RETURN_LANG is incorrect in the case when x is longer than LANG_MAXLEN, because strncpy won't add a terminating null in that case, resulting in a non-null-terminated string. 

(I realize this a pre-existing issue but we might want to consider whether it needs to be fixed because it might lead to a varnish crash.)

From http://man7.org/linux/man-pages/man3/strcpy.3.html:  "If there is no null byte among the first n  bytes of src, the string placed in dest will not be null-terminated."

wjordan replied:
If this code is safe as long as x is not longer than `LANG_MAXLEN` (8), then it should be safe as long as none of the keys in our [list of whitelisted languages](https://github.com/code-dot-org/code-dot-org/blob/staging/cookbooks/cdo-varnish/files/default/languages.json) exceeds this length (max is currently 5), does that sound right?

philbogle replied:
We can't be certain that list won't change or that someone won't use this function in some new way.

It's easy to be safe in all cases by unconditionally adding the null termination ourselves after the strncpy, which is standard practice:
strncpy(lang, x, LANG_MAXLEN);  \
lang[LANG_MAXLEN - 1]  = '\0' \

philbogle commented on cookbooks/cdo-varnish/templates/default/accept-language.vcl.erb:
```diff
@@ -0,0 +1,188 @@
+        PUSH_LANG(lang_tok, q);
+
+        /* For cases like 'en-GB', we also want the root language in the final list */
+        if ('-' == lang_tok[2]) {
```
I believe this will dereference off the end of the string and potentially crash varnish if strlen(lang_tok) < 2-.  They don't check the length of the string and unconditionally reference lang_tok[2[.

philbogle commented on cookbooks/cdo-varnish/templates/default/accept-language.vcl.erb:
```diff
@@ -0,0 +1,188 @@
+            root_lang[0] = lang_tok[0];
+            root_lang[1] = lang_tok[1];
+            root_lang[2] = '\0';
+            PUSH_LANG(root_lang, q - 0.001);
```
If lang_tok equals (say "*n-gb", then this will result in an 0 - 0.001 which is an invalid negative q number.

philbogle commented on cookbooks/cdo-varnish/templates/default/default.vcl.erb:
```diff
@@ -555,53 +154,20 @@ sub vcl_deliver {
-    set obj.status = 302;
+# Handle redirects
+sub vcl_synth {
+  if(resp.status == 750) {
```
Could you add a comment explaining what 750 means?

philbogle commented on cookbooks/cdo-varnish/test/integration/default/minitest/test_default.rb:
```diff
@@ -0,0 +1,293 @@
+    text_en = 'Hello World!'
+    text_fr = 'Bonjour le Monde!'
+    mock_url(url, text_en, {'X-Varnish-Accept-Language' => 'en'}, {vary: 'X-Varnish-Accept-Language'})
+    mock_url(url, text_fr, {'X-Varnish-Accept-Language' => 'fr'}, {vary: 'X-Varnish-Accept-Language'})
```
Would be curious to see if you can repro any bad behavior where Accept-Language = 'f' (i.e. less than two characters.) Odds of a crash are small of course.

wjordan replied:
Added a passing test (0e20ad781d9ae9fa6cdac2c67cc94521ff7c77c8) for weird Accept-Language request headers. (I ran it over 5000 iterations once without any failures as well). Further testing of any crash-prone edge cases would probably require a more elaborate memory-corruption harness (e.g., [varnish-valgrind](https://github.com/lkarsten/varnish-valgrind) or something similar)

philbogle commented on lib/cdo/aws/cloudfront.rb:
```diff
@@ -0,0 +1,211 @@
+require_relative '../../../deployment'
```
Does this config end up getting tested?  I see where we do this for varnish (in cdo-varnish/test/integration/default/minitest/test_default.rb) but not a corresponding test for cloudfront.

wjordan replied:
Added CloudFront unit tests to [test_cloudfront.rb](https://github.com/code-dot-org/code-dot-org/blob/b781b19ee8812d56eaeaeb0306d784772571fdb1/shared/test/test_cloudfront.rb), with the following note:

> These unit tests simply confirm that the #create_or_update method will pass
> properly-structured data to the AWS CloudFront Client library.
> Separate integration tests are required to guarantee that the live API
> endpoints will accept the values provided.

I think manual testing of the CloudFront distribution created by this script should be fine for now, and we can look into adding integration tests later if possible.

philbogle replied:
My take:  Manual testing before initial deployment is OK, but we will need
to add automated testing to verify that the CDN is behaving as expected
before we can consider the CDN work done.  This isn't an area where we want
to incur technical debt because it's critical for both scale and
correctness , HOC is coming up very soon, and because development doesn't
give us good coverage for what the CDN is doing.

Is there any way we can assertions about caching headers to our existing UI
tests that go through the CDN to see if the CDN is behaving as expected?

On Mon, Oct 5, 2015 at 9:19 AM, Will Jordan notifications@github.com
wrote:

> In lib/cdo/aws/cloudfront.rb
> https://github.com/code-dot-org/code-dot-org/pull/4207#discussion_r41164020
> :
> 
> > @@ -0,0 +1,211 @@
> > +require_relative '../../../deployment'
> 
> Added CloudFront unit tests to test_cloudfront.rb
> https://github.com/code-dot-org/code-dot-org/blob/b781b19ee8812d56eaeaeb0306d784772571fdb1/shared/test/test_cloudfront.rb,
> with the following note:
> 
> These unit tests simply confirm that the #create_or_update method will pass
> properly-structured data to the AWS CloudFront Client library.
> Separate integration tests are required to guarantee that the live API
> endpoints will accept the values provided.
> 
> I think manual testing of the CloudFront distribution created by this
> script should be fine for now, and we can look into adding integration
> tests later if possible.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/code-dot-org/code-dot-org/pull/4207/files#r41164020.

wjordan replied:
Finished setting up a basic suite of CloudFront integration tests matching the existing Varnish test, please see #4440.

philbogle commented on lib/cdo/aws/cloudfront.rb:
```diff
@@ -0,0 +1,211 @@
+      cloudfront = CONFIG[app]
+      behaviors = config[:behaviors].map do |behavior|
+        if behavior[:path].is_a? Array
+          behavior[:path].map do |path|
```
Could you refactor to avoid duplicating the code for the array and non-array case?  (e.g. wrap in an array if not already an array)

philbogle commented on lib/cdo/aws/cloudfront.rb:
```diff
@@ -0,0 +1,211 @@
+        default_cache_behavior: cache_behavior(config[:default]),
+        cache_behaviors: {
+          quantity: behaviors.length, # required
+          items: behaviors.empty? ? nil : behaviors,
```
Do we want this to be nil if behaviors are empty, or not present at all?

wjordan replied:
I think trial and error showed that the CloudFront API would only accept this very particular format for certain required empty lists.

philbogle commented on lib/cdo/aws/cloudfront.rb:
```diff
@@ -0,0 +1,211 @@
+          },
+        },
+      }.tap do |cf|
+        cf[:caller_reference] = reference || Digest::MD5.hexdigest(Marshal.dump(config)) # required
```
How is this used?  Will anything bad happen if the implementation for Marshal changes so that the same object is marhselled differently?

wjordan replied:
Forgot to follow up on this before. Here's the [documentation](http://docs.aws.amazon.com/sdkforruby/api/Aws/CloudFront/Types/DistributionConfig.html#caller_reference-instance_method) for `caller_reference`:

> A unique number that ensures the request can't be replayed. If the CallerReference is new (no matter the content of the DistributionConfig object), a new distribution is created. If the CallerReference is a value you already sent in a previous request to create a distribution, and the content of the DistributionConfig is identical to the original request (ignoring white space), the response includes the same information returned to the original request. If the CallerReference is a value you already sent in a previous request to create a distribution but the content of the DistributionConfig is different from the original request, CloudFront returns a DistributionAlreadyExists error.

As I understand this the caller_reference is only used to prevent replayed requests from causing duplicated side-effects anyway, so there's little-to-no likelihood of something bad happening if the Marshal implementation changes.

philbogle commented on shared/test/test_varnish_helpers.rb:
```diff
@@ -0,0 +1,130 @@
+      '*/images',
+      # Wildcard at both ends
+      '*images*',
+      # Maximum path length
```
Could you add checks for characters like [] and ()?  These currently aren't being escape in regular expressions so it's important they don't occur in paths.

wjordan replied:
added in 1a507d21da4bf1a2881c8ba33e45fc7a96fbabe8.

philbogle replied:
yes, I wouldn't insist that you to fix all of this in the current PR; I
just wanted to invest the time now to look at in detail and understand the
issues while I had a chance to look at it.

I will create a tracking bug that summarizes all of the issues.

On Fri, Oct 2, 2015 at 5:30 PM, Will Jordan notifications@github.com
wrote:

> In shared/test/test_varnish_helpers.rb
> https://github.com/code-dot-org/code-dot-org/pull/4207#discussion_r41080914
> :
> 
> > +
> > -  def test_invalid_path_patterns
> > -    [
> > -      # Unsupported extension wildcards
> > -      '/images/*.jpg',
> > -      '/a??.jpg',
> > -      '/a*.jpg',
> > -      # Wildcard-extension
> > -      '/_.doc_',
> > -      # Path with wildcard subexpression
> > -      '/assets/*/images',
> > -      # Extension-wildcard containing path delimiter
> > -      '*/images',
> > -      # Wildcard at both ends
> > -      '_images_',
> > -      # Maximum path length
> 
> added in 1a507d2
> https://github.com/code-dot-org/code-dot-org/commit/1a507d21da4bf1a2881c8ba33e45fc7a96fbabe8
> .
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/code-dot-org/code-dot-org/pull/4207/files#r41080914.

philbogle replied:
Is there any way we can incorporate those tests in our build in case we
change this?  Do you happen to know how they run them?

On Fri, Oct 2, 2015 at 5:38 PM, Phil Bogle phil@code.org wrote:

> yes, I wouldn't insist that you to fix all of this in the current PR; I
> just wanted to invest the time now to look at in detail and understand the
> issues while I had a chance to look at it.
> 
> I will create a tracking bug that summarizes all of the issues.
> 
> On Fri, Oct 2, 2015 at 5:30 PM, Will Jordan notifications@github.com
> wrote:
> 
> > In shared/test/test_varnish_helpers.rb
> > https://github.com/code-dot-org/code-dot-org/pull/4207#discussion_r41080914
> > :
> > 
> > > +
> > > -  def test_invalid_path_patterns
> > > -    [
> > > -      # Unsupported extension wildcards
> > > -      '/images/*.jpg',
> > > -      '/a??.jpg',
> > > -      '/a*.jpg',
> > > -      # Wildcard-extension
> > > -      '/_.doc_',
> > > -      # Path with wildcard subexpression
> > > -      '/assets/*/images',
> > > -      # Extension-wildcard containing path delimiter
> > > -      '*/images',
> > > -      # Wildcard at both ends
> > > -      '_images_',
> > > -      # Maximum path length
> > 
> > added in 1a507d2
> > https://github.com/code-dot-org/code-dot-org/commit/1a507d21da4bf1a2881c8ba33e45fc7a96fbabe8
> > .
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/code-dot-org/code-dot-org/pull/4207/files#r41080914.

philbogle replied:
(Again, not needed right away, I will track this in the bug.)

On Fri, Oct 2, 2015 at 5:39 PM, Phil Bogle phil@code.org wrote:

> Is there any way we can incorporate those tests in our build in case we
> change this?  Do you happen to know how they run them?
> 
> On Fri, Oct 2, 2015 at 5:38 PM, Phil Bogle phil@code.org wrote:
> 
> > yes, I wouldn't insist that you to fix all of this in the current PR; I
> > just wanted to invest the time now to look at in detail and understand the
> > issues while I had a chance to look at it.
> > 
> > I will create a tracking bug that summarizes all of the issues.
> > 
> > On Fri, Oct 2, 2015 at 5:30 PM, Will Jordan notifications@github.com
> > wrote:
> > 
> > > In shared/test/test_varnish_helpers.rb
> > > https://github.com/code-dot-org/code-dot-org/pull/4207#discussion_r41080914
> > > :
> > > 
> > > > +
> > > > -  def test_invalid_path_patterns
> > > > -    [
> > > > -      # Unsupported extension wildcards
> > > > -      '/images/*.jpg',
> > > > -      '/a??.jpg',
> > > > -      '/a*.jpg',
> > > > -      # Wildcard-extension
> > > > -      '/_.doc_',
> > > > -      # Path with wildcard subexpression
> > > > -      '/assets/*/images',
> > > > -      # Extension-wildcard containing path delimiter
> > > > -      '*/images',
> > > > -      # Wildcard at both ends
> > > > -      '_images_',
> > > > -      # Maximum path length
> > > 
> > > added in 1a507d2
> > > https://github.com/code-dot-org/code-dot-org/commit/1a507d21da4bf1a2881c8ba33e45fc7a96fbabe8
> > > .
> > > 
> > > —
> > > Reply to this email directly or view it on GitHub
> > > https://github.com/code-dot-org/code-dot-org/pull/4207/files#r41080914
> > > .

philbogle replied:
Tracking bug is here:   https://www.pivotaltracker.com/story/show/104775628

This will become moot, of course, if we no longer rely on Varnish or at
least simplify the config so that it's not doing language header
simplification in C code.  Presumably the same issue happens in the CDN so
I assume the CDN has a solution for this, how does that work for most uers
who aren't using Varnish?

On Fri, Oct 2, 2015 at 5:43 PM, Phil Bogle phil@code.org wrote:

> (Again, not needed right away, I will track this in the bug.)
> 
> On Fri, Oct 2, 2015 at 5:39 PM, Phil Bogle phil@code.org wrote:
> 
> > Is there any way we can incorporate those tests in our build in case we
> > change this?  Do you happen to know how they run them?
> > 
> > On Fri, Oct 2, 2015 at 5:38 PM, Phil Bogle phil@code.org wrote:
> > 
> > > yes, I wouldn't insist that you to fix all of this in the current PR; I
> > > just wanted to invest the time now to look at in detail and understand the
> > > issues while I had a chance to look at it.
> > > 
> > > I will create a tracking bug that summarizes all of the issues.
> > > 
> > > On Fri, Oct 2, 2015 at 5:30 PM, Will Jordan notifications@github.com
> > > wrote:
> > > 
> > > > In shared/test/test_varnish_helpers.rb
> > > > https://github.com/code-dot-org/code-dot-org/pull/4207#discussion_r41080914
> > > > :
> > > > 
> > > > > +
> > > > > -  def test_invalid_path_patterns
> > > > > -    [
> > > > > -      # Unsupported extension wildcards
> > > > > -      '/images/*.jpg',
> > > > > -      '/a??.jpg',
> > > > > -      '/a*.jpg',
> > > > > -      # Wildcard-extension
> > > > > -      '/_.doc_',
> > > > > -      # Path with wildcard subexpression
> > > > > -      '/assets/*/images',
> > > > > -      # Extension-wildcard containing path delimiter
> > > > > -      '*/images',
> > > > > -      # Wildcard at both ends
> > > > > -      '_images_',
> > > > > -      # Maximum path length
> > > > 
> > > > added in 1a507d2
> > > > https://github.com/code-dot-org/code-dot-org/commit/1a507d21da4bf1a2881c8ba33e45fc7a96fbabe8
> > > > .
> > > > 
> > > > —
> > > > Reply to this email directly or view it on GitHub
> > > > https://github.com/code-dot-org/code-dot-org/pull/4207/files#r41080914
> > > > .

philbogle commented:
Done with main code review pass, thank you for the added tests!

philbogle commented:
LGTM after comments and merge conflicts are addressed.

philbogle commented on cookbooks/cdo-varnish/templates/default/accept-language.vcl.erb:
```diff
@@ -0,0 +1,188 @@
+    /* Search ":<lang>:" in supported languages string */
+    strncpy(match_str, ":", 1);
+    strncat(match_str, lang, LANG_MAXLEN);
+    strncat(match_str, ":\0", 2);
```
I believe this is another potential overflow.

strncat, unlike strncpy, always appends a terminate null character.  So the maximum number of characters copies is LANG_MAXLEN +4

philbogle commented on cookbooks/cdo-varnish/templates/default/accept-language.vcl.erb:
```diff
@@ -0,0 +1,188 @@
+#define PUSH_LANG(x,y) { \
+    /* fprintf(stderr, "Pushing lang [%d] %s %.4f\n", curr_lang, x, y); */ \
+    /* We have to copy, otherwise root_lang will be the same every time */ \
+    strncpy(pl[curr_lang].lang, x, LANG_MAXLEN); \
```
in this case, x is derived from an HTTP header so it can be arbitrarily long.

This means than pl[curr_lang].lang will lack the terminating null so calling string functions will reference off the end of the string.

wjordan commented:
@philbogle thanks, I'll follow up on the remaining comments and then let you know.

Regarding the comments on `accept-language.vcl.erb`, I think we would best follow up on those upstream at [varnish-accept-language](https://github.com/cosimo/varnish-accept-language), as I'm not knowledgeable enough about that low-level library to respond directly myself. Perhaps a look at the upstream [tests](https://github.com/cosimo/varnish-accept-language/tree/4.0/t) will satisfy some of your concerns, or we could suggest additional tests to be added there (especially if there are failing tests we discover).

I think an upstream security-audit of this code would be valuable if we ever find the time for it, but I think it's also a quite separate task from the present work. The code was not substantially modified by this PR (I've merely extracted it into a separate file in order to ensure it will continue to be untouched), and it remains unchanged from the upstream codebase, so I think it's out of scope of this particular code review.

wjordan commented:
Rebase is failing tests because #4267 (session key refactoring which this branch depends on) was reverted in #4293 due to some session-cookie UI test failure issue that was not yet resolved.

philbogle commented:
Clearing myself as assignee, please re-add me if you'd like me to take another look at this PR.

wjordan commented:
Fully rebased, will merge once all tests pass.

wjordan commented:
circleci failures are unrelated to this PR.
